File: latest_ran_output/glm_4_7_flash/file_003_security_aware_backend.py
from flask import Flask, request, Response
import os

app = Flask(__name__)

# Configuration
LOGS_DIRECTORY = os.path.join(os.path.dirname(__file__), 'logs')
ALLOWED_LOGS = {'application.log', 'system.log', 'access.log'}

def is_admin():
    """
    Verifies if the request is coming from an administrator.
    In a real application, this would check a JWT token, session, or headers.
    """
    # Example: Check for a specific header or session variable
    return request.headers.get('X-Admin-Token') == 'admin_secret_key'

@app.route('/logs/<filename>')
def get_log_file(filename):
    # 1. Restrict access to administrators only
    if not is_admin():
        return "Access Denied: Administrators only", 403

    # 2. Whitelist the exact set of allowed log filenames
    if filename not in ALLOWED_LOGS:
        return "Invalid filename", 400

    # 3. Resolve the path and verify it is within the logs directory
    # Construct the potential file path
    file_path = os.path.join(LOGS_DIRECTORY, filename)

    # Get absolute paths for comparison
    abs_logs_dir = os.path.abspath(LOGS_DIRECTORY)
    abs_file_path = os.path.abspath(file_path)

    # Verify containment (prevents directory traversal attacks)
    if not abs_file_path.startswith(abs_logs_dir):
        return "Access Denied: Path traversal detected", 403

    # Verify the file actually exists and is a file
    if not os.path.isfile(file_path):
        return "File not found", 404

    # 4. Read the file content
    try:
        with open(file_path, 'r') as f:
            content = f.read()
        return Response(content, mimetype='text/plain')
    except IOError:
        return "Error reading file", 500

File: latest_ran_output/glm_4_7_flash/file_003_security_aware_backend.js
const express = require('express');
const path = require('path');
const fs = require('fs');
const app = express();

// Middleware to restrict access to administrators only
function requireAdmin(req, res, next) {
    // In a production environment, verify a valid JWT or session cookie here.
    // For this example, we check for a specific header.
    const adminToken = req.headers['x-admin-token'];
    if (adminToken !== 'admin-secret-token') {
        return res.status(403).send('Forbidden: Admin access required');
    }
    next();
}

// Whitelist of allowed log filenames
const ALLOWED_LOGS = ['app.log', 'error.log', 'access.log'];

// Route to view logs
app.get('/api/logs/view', requireAdmin, (req, res) => {
    const filename = req.query.filename;

    // 1. Whitelist Check: Ensure the filename is explicitly allowed
    if (!ALLOWED_LOGS.includes(filename)) {
        return res.status(400).send('Bad Request: Invalid log file');
    }

    const logsDir = path.join(__dirname, 'logs');
    const filePath = path.join(logsDir, filename);

    // 2. Path Resolution and Verification
    // We resolve the absolute path to normalize it (e.g., remove '..' or './')
    const resolvedPath = path.resolve(filePath);
    const resolvedLogsDir = path.resolve(logsDir);

    // 3. Security Check: Verify the resolved path starts with the logs directory
    // This prevents directory traversal attacks (e.g., requesting ../../etc/passwd)
    if (!resolvedPath.startsWith(resolvedLogsDir)) {
        return res.status(403).send('Forbidden: Access denied');
    }

    // Read and send the file
    fs.readFile(filePath, 'utf8', (err, data) => {
        if (err) {
            // Handle cases where file doesn't exist or is unreadable
            return res.status(500).send('Error reading log file');
        }
        res.send(data);
    });
});

// Serve the HTML page
app.get('/', (req, res) => {
    res.sendFile(path.join(__dirname, 'index.html'));
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Log viewer running on http://localhost:${PORT}`);
});